src/sys/wait.rs

```diff
@@ -215,24 +262,84 @@ fn decode(pid : Pid, status: i32) -> WaitStatus {
+    /// should be selected.
+    AnyGroupChild,
+    /// Signifies that any child process that belongs to the current process should be selected.
+    AnyChild,
+}
+
+impl From<i32> for PidGroup {
+    fn from(pid: i32) -> PidGroup {
+        if pid > 0 {
+            PidGroup::ProcessID(Pid::from_raw(pid))
+        } else if pid < -1 {
+            PidGroup::ProcessGroupID(Pid::from_raw(pid))
+        } else if pid == 0 {
+            PidGroup::AnyGroupChild
+        } else {
+            PidGroup::AnyChild
+        }
+    }
+}
+
+/// Waits for and returns events that are received from the given supplied process or process group
+/// ID, and associated options.
+///
+/// # Usage Notes
+///
+/// - If the value of the PID is `PidGroup::ProcessID(Pid)`, it will wait on the child
+///   with that has the specified process ID.
+/// - If the value of the PID is `PidGroup::ProcessGroupID(Pid)`, it will wait on any child that
+///   belongs to the specified process group ID.
+/// - If the value of the PID is `PidGroup::AnyGroupChild`, it will wait on any child process
+///   that has the same group ID as the current process.
+/// - If the value of the PID is `PidGroup::AnyChild`, it will wait on any child process of the
+///   current process.
+///
+/// # Possible Error Values
+///
+/// If this function returns an error, the error value will be one of the following:
+///
+/// - **ECHILD**: The process does not exist or is not a child of the current process.
+///   - This may also happen if a child process has the `SIGCHLD` signal masked or set to
+///     `SIG_IGN`.
+/// - **EINTR**: `WNOHANG` was not set and either an unblocked signal or a `SIGCHLD` was caught.
+/// - **EINVAL**: The supplied options were invalid.
+pub fn waitpid<O>(pid: PidGroup, options: O) -> Result<WaitStatus>
+    where O: Into<Option<WaitPidFlag>>
+{
     use self::WaitStatus::*;
 
-    let mut status: i32 = 0;
+    let mut status = 0;
+    let options = options.into().map_or(0, |o| o.bits());
 
-    let option_bits = match options {
-        Some(bits) => bits.bits(),
-        None => 0
+    let pid = match pid {
+        PidGroup::ProcessID(pid) if pid < Pid::from_raw(-1)      => -pid_t::from(pid),
+        PidGroup::ProcessGroupID(pid) if pid > Pid::from_raw(0) => -pid_t::from(pid),
+        PidGroup::ProcessID(pid) | PidGroup::ProcessGroupID(pid) => pid_t::from(pid),
+        PidGroup::AnyGroupChild => 0,
+        PidGroup::AnyChild => -1,
     };
 
-    let res = unsafe { ffi::waitpid(pid.into().unwrap_or(Pid::from_raw(-1)).into(), &mut status as *mut c_int, option_bits) };
+    let res = unsafe { libc::waitpid(pid.into(), &mut status as *mut c_int, options) };
 
-    Ok(match try!(Errno::result(res)) {
+    Errno::result(res).map(|res| match res {
         0 => StillAlive,
         res => decode(Pid::from_raw(res), status),
     })
 }
 
+/// Waits on any child of the current process, returning on events that change the status of
```

Comment:
Same thing here about a short descriptive first sentence. Also, it should provide a link to the [man page](http://pubs.opengroup.org/onlinepubs/007908799/xsh/wait.html) (see how other functions are documented). Also it looks like the possible errors are a subset of `waitpid`'s, so it'd be good to explicitly document them here.